### PR TITLE
Select group when adding new credentials

### PR DESCRIPTION
**Pull request by varjolintu**
Allows selecting a group when adding new credentials to KeePassXC from the popup dialog.

A new command `get-database-groups` is added to the protocol. If older version than KeePassXC 2.4.0 is used, credentials are saved to the default KeePassXC-Browser Passwords group. For saving, both name and group UUID are used.

The groups are read from a dynamic array where each group can contain N children:
```
{
    name: '<group name>',
    uuid: '<group UUID>',
    children: [...]
}
```

Fixes https://github.com/keepassxreboot/keepassxc-browser/issues/382.
Fixes https://github.com/keepassxreboot/keepassxc-browser/issues/388.
Fixes https://github.com/keepassxreboot/keepassxc-browser/issues/427.

Related KeePassXC PR: https://github.com/keepassxreboot/keepassxc/pull/2637.

groups_keepassxc

groups_popup

Also, a default group can be set, or the group can be asked every time:
Screen Shot 2019-03-09 at 10 32 40

### CODE REVIEW

**Comment by varjolintu** [comment]
I updated the PR to include a tree view.

**Comment by droidmonkey** [comment]
Is the tree collapsed by default? (Should be) the arrows don't seem to indicate the correct state.

**Comment by varjolintu** [comment]
@droidmonkey It's not collapsed, or cannot be collapsed. It's just a tree view. What do you suggest instead of the arrows? Or should it definitely be able to collapse?

**Comment by varjolintu** [comment]
I changed the icon to arrow down, indicating open tree. For now.

File: keepassxc-browser/background/keepass.js [hunk 1]
```diff
@@ -696,6 +702,71 @@ keepass.lockDatabase = function(tab) {
...
+                const res = keepass.decrypt(response.message, response.nonce);
+                if (!res) {
+                    keepass.handleError(tab, kpErrors.CANNOT_DECRYPT_MESSAGE);
+                    callback([]);
+                    return;
+                }
+
+                const message = nacl.util.encodeUTF8(res);
+                const parsed = JSON.parse(message);
+                keepass.setcurrentKeePassXCVersion(parsed.version);
```
Comment (droidmonkey) [inline keepassxc-browser/background/keepass.js]: this seems out of place

File: keepassxc-browser/background/keepass.js [hunk 2]
```diff
@@ -696,6 +702,71 @@ keepass.lockDatabase = function(tab) {
...
+
+                if (keepass.verifyResponse(parsed, incrementedNonce)) {
+                    groups = parsed.groups;
+                    keepass.updateLastUsed(keepass.databaseHash);
+                    callback(groups);
+                } else {
+                    console.log('getDatabaseGroups rejected');
```
Comment (droidmonkey) [inline keepassxc-browser/background/keepass.js]: Would you want to show a proper error at this point? Would the CANNOT_DECRYPT_MESSAGE be appropriate here?

Comment (droidmonkey) [reply keepassxc-browser/background/keepass.js]: Something like "Invalid or corrupt message sent from KeePassXC"

Comment (varjolintu) [reply keepassxc-browser/background/keepass.js]: Something like that yes. Currently the verifyResponse() returns false from multiple reasons, so the message must be a general one. The one you proposed or something like: "Message sent from KeePassXC cannot be verified".

**Comment by varjolintu** [comment]
I added one more thing to the PR. A default group can be specified, or the user can set an option that asks for the group every time when saving new credentials.

Screen Shot 2019-03-09 at 10 32 40

**Comment by droidmonkey** [comment]
There is still a problem. When defining the group to save to, if it does not exist then the entry is silently dropped. The group should be created to store the new entry if it doesn't already exist. This will likely require changes on KeePassXC side.

**Comment by varjolintu** [comment]
It also seems that triggering notifications from the popup doesn't work. Some changes are needed for that too. Probably we need to call the background script to activate them.

**Comment by varjolintu** [comment]
@droidmonkey That is now done. KeePassXC side support is here: https://github.com/keepassxreboot/keepassxc/pull/2790.

File: keepassxc-browser/_locales/en/messages.json [hunk 3]
```diff
@@ -312,9 +312,13 @@
         "description": "A popup message shown choosing what group user wants to use for new credentials."
     },
     "popupRememberErrorDefaultGroupNotFound": {
-        "message": "Error: Specified default group not found.",
+        "message": "Error: Specified default group not found. Creating all necessary groups.",
```
Comment (droidmonkey) [inline keepassxc-browser/_locales/en/messages.json]: This is not an error in this case, just an info. Should be reworded as: "Info: Default group created in connected database"

File: keepassxc-browser/_locales/en/messages.json [hunk 4]
```diff
@@ -312,9 +312,13 @@
...
         "description": "Error message shown when default group set cannot be found."
     },
+    "popupRememberErrorCreatingNewGroup": {
+        "message": "New group(s) cannot be created.",
```
Comment (droidmonkey) [inline keepassxc-browser/_locales/en/messages.json]: "Error: .... "

**Comment by droidmonkey** [comment]
I thought of a little vulnerability with create entry/group. There is really nothing stopping abuse of the API to create an entry or group. We prevent retrieving information from the database by showing allow/deny dialog, but we do not prevent writing to the database (from KeePassXC).

**Comment by varjolintu** [comment]
@droidmonkey Yes. We need to restrict this in some way.

**Comment by droidmonkey** [comment]
We can introduce the protections in 2.4.1

**Pull request merged**